outer.c

```diff
@@ -279,36 +304,88 @@ struct outer_helper outer_helper_spawn(void)
 	id_str = itoa(gid);
 	populate_id_map(map, sizeof (map), "/etc/subgid", id_str, group ? group->gr_name : id_str);
 	burn(procfd, "gid_map", map);
+}
 
-	/* Notify sibling that we're done changing their [ug]id map */
-	int ok = 1;
-	write(pipefds_in[1], &ok, sizeof (ok));
+struct persistflag {
+	int flag;
+	const char *proc_ns_name;
+};
 
-	_exit(0);
+static struct persistflag pflags[] = {
+	{ BST_CLONE_NEWCGROUP, "cgroup" },
+	{ BST_CLONE_NEWIPC,    "ipc",   },
+	{ BST_CLONE_NEWNS,     "mnt"    },
+	{ BST_CLONE_NEWNET,    "net"    },
+	{ BST_CLONE_NEWPID,    "pid"    },
+	{ BST_CLONE_NEWTIME,   "time"   },
+	{ BST_CLONE_NEWUSER,   "user"   },
+	{ BST_CLONE_NEWUTS,    "uts"    },
+};
+
+static void persist_ns_files(int pid, const char *persist) {
+	char procname[PATH_MAX];
+	snprintf(procname, sizeof(procname), "/proc/%d/ns", pid);
+	int procnsdir = open(procname, O_DIRECTORY | O_PATH);
+	if (procnsdir < 0) {
+		err(1, "open(\"%s\")", procname);
+	}
+	int persistdir = open(persist, O_DIRECTORY | O_PATH);
+	if (persistdir < 0) {
+		err(1, "open(\"%s\")", persist);
+	}
+	for (struct persistflag *f = pflags; f < pflags + lengthof(pflags); f++) {
+		// Need permission check here?
+		int nsfd = openat(persistdir, f->proc_ns_name, O_CREAT | O_WRONLY | O_EXCL, 0666);
+		if (nsfd < 0) {
+			if (errno != EEXIST) {
+				err(1, "creat(\"%s/%s\")", persist, f->proc_ns_name);
+			}
+		} else {
+			close(nsfd);
+		}
+		// Where is mountat()?  Thankfully, we can still name the persist directory.
+		char procname1[PATH_MAX];
+		snprintf(procname1, sizeof(procname1), "/proc/%d/ns/%s", pid, f->proc_ns_name);
```

Comment:
You should be able to just redefine procname here

Reply:
I actually just re-used the same procname local.